sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -794,17 +794,25 @@ static const bool FilterEventWaitList = [] {
 }();
 
 pi_result _pi_ze_event_list_t::createAndRetainPiZeEventList(
-    pi_uint32 EventListLength, const pi_event *EventList, pi_queue CurQueue) {
+    pi_uint32 EventListLength, const pi_event *EventList,
+    ze_command_list_handle_t ZeCommandList, pi_queue CurQueue) {
   this->Length = 0;
   this->ZeEventList = nullptr;
   this->PiEventList = nullptr;
 
-  if (EventListLength > 0) {
-    try {
+  try {
+    if (CurQueue->InorderQueue &&
+        ZeCommandList == (*CurQueue->PreviousEvent)->ZeCommandList) {
```

Comment:
With the changes coming in https://github.com/intel/llvm/pull/3369 commands are going to be submitted to two different queues. For that I think we should always draw the dependency irrelevant of the command-list. It might be some overhead for OOO queues, but let's try to keep it simple for now (maybe add a TODO comment here about the possibility to avoid the extra dependency).

Reply:
my concern here is the last command (L) of a command-list and the first command (F) of a new command-list. as CurQueue->PreviousEven (which is for command L) should not be added to the waitlist of command F. adding this dependency does not invalidate anything though.

Reply:
> CurQueue->PreviousEven (which is for command L) should not be added to the waitlist of command F

And I am saying let's add it too.

Command-lists on **the same** L0 queue are executed in-order, but with multiple L0 queues coming in https://github.com/intel/llvm/pull/3369 it's not so. Thus a command-list from one queue competes with the command-lists from another queue. To gain the in-order semantics we need to draw the dependency between *every* adjacent commands, I think.